src/trustRegion.jl

```diff
@@ -402,12 +418,12 @@ function perform_step!(cache::TrustRegionCache{true})
     @unpack make_new_J, J, fu, f, u, p, u_tmp, alg, linsolve = cache
     if cache.make_new_J
         jacobian!(J, cache)
-        mul!(cache.H, J, J)
-        mul!(cache.g, J, fu)
+        mul!(cache.H, J', J)
+        mul!(cache.g, J', fu)
         cache.stats.njacs += 1
     end
 
-    linres = dolinsolve(alg.precs, linsolve, A = cache.H, b = _vec(cache.g),
+    linres = dolinsolve(alg.precs, linsolve, A = J, b = _vec(fu), # cache.H, b = _vec(cache.g),
```

Comment:
I think this is incorrect because here $J$ is the jacobian of the actual function and not the merit function $0.5 * ||f(x)||_{2}^2$, but the initial estimate of the step size `u_tmp` is actually as calculated $-H^{-1}g$, where H and g are the hessian and gradient of the merit func at `u` ?

Reply:
That is a subtle point. You are right that on paper you do a Gauss-Newton step, i.e., $p_{GN} = -(J^\top J)^{-1}J^\top f$ (since we use $H = J^\top J$ as proxy for the Hessian of the merit function and $g= J^\top f$). But clearly, if $J$ is non-singular this is equivalent to just the Newton step, i.e., $p_{GN} = p_N = -J^{-1} f$. At the same time, computing $J^\top J$ squares the condition number of $J$ which makes using the Gauss-Newton step unfavorable. In `NLsolve`, the same is done actually. The only difference is that they default to computing the Moore-Penrose inverse of $J$ via the SVD when it is singular.  We don't have that exception yet (unless it is burried in `dolinsolve`?). I meant to ask how we should go about this actually. I feel like this should be handled by `LinearSolve` via some option which is why I didn't put it in here yet.

Reply:
@ChrisRackauckas  Any thoughts on how to go about the case of singular Jacobians? Is this already handled in `LinearSolve`, for example triggered by an LAPACK singularity error or so? If not, I feel it really should go there as opposed to here. That said, putting it here would be a quick solution for now.

Reply:
LinearSolve generally attempts to be non-blocking, so it will return a solution without erroring, though the return code may indicate that the solution did not solve correctly. It should just be a retcode check then, though specifically with singularity issues we probably need to add a few retcode improvements. In the ODE solver that is generally not checked because the adaptivity handles it appropriately, though this is a general improvement we should do anyways.